registry/storage/driver/testsuites/testsuites.go

```diff
@@ -424,180 +418,180 @@ func (suite *DriverSuite) testContinueStreamAppend(c *check.C, chunkSize int64)
+	suite.Require().Equal(int64(len(fullContents[curSize:])), nn)
 
 	err = writer.Commit(context.Background())
-	c.Assert(err, check.IsNil)
+	suite.Require().NoError(err)
 	err = writer.Close()
-	c.Assert(err, check.IsNil)
+	suite.Require().NoError(err)
 
 	received, err := suite.StorageDriver.GetContent(suite.ctx, filename)
-	c.Assert(err, check.IsNil)
-	c.Assert(received, check.DeepEquals, fullContents)
+	suite.Require().NoError(err)
+	suite.Require().Equal(fullContents, received)
 }
 
 // TestReadNonexistentStream tests that reading a stream for a nonexistent path
 // fails.
-func (suite *DriverSuite) TestReadNonexistentStream(c *check.C) {
+func (suite *DriverSuite) TestReadNonexistentStream() {
 	filename := randomPath(32)
 
 	_, err := suite.StorageDriver.Reader(suite.ctx, filename, 0)
-	c.Assert(err, check.NotNil)
-	c.Assert(err, check.FitsTypeOf, storagedriver.PathNotFoundError{})
-	c.Assert(strings.Contains(err.Error(), suite.Name()), check.Equals, true)
+	suite.Require().Error(err)
+	suite.Require().IsType(err, storagedriver.PathNotFoundError{})
+	suite.Require().Contains(err.Error(), suite.Name())
 
 	_, err = suite.StorageDriver.Reader(suite.ctx, filename, 64)
-	c.Assert(err, check.NotNil)
-	c.Assert(err, check.FitsTypeOf, storagedriver.PathNotFoundError{})
-	c.Assert(strings.Contains(err.Error(), suite.Name()), check.Equals, true)
+	suite.Require().Error(err)
+	suite.Require().IsType(err, storagedriver.PathNotFoundError{})
+	suite.Require().Contains(err.Error(), suite.Name())
 }
 
 // TestWriteZeroByteStreamThenAppend tests if zero byte file handling works for append to a Stream
-func (suite *DriverSuite) TestWriteZeroByteStreamThenAppend(c *check.C) {
+func (suite *DriverSuite) TestWriteZeroByteStreamThenAppend() {
 	filename := randomPath(32)
-	defer suite.deletePath(c, firstPart(filename))
+	defer suite.deletePath(firstPart(filename))
 	chunkSize := int64(32)
 	contentsChunk1 := randomContents(chunkSize)
 
 	// Open a Writer
 	writer, err := suite.StorageDriver.Writer(suite.ctx, filename, false)
-	c.Assert(err, check.IsNil)
+	suite.Require().NoError(err)
 
 	// Close the Writer
 	err = writer.Commit(context.Background())
-	c.Assert(err, check.IsNil)
+	suite.Require().NoError(err)
 	err = writer.Close()
-	c.Assert(err, check.IsNil)
+	suite.Require().NoError(err)
 	curSize := writer.Size()
-	c.Assert(curSize, check.Equals, int64(0))
+	suite.Require().Equal(int64(0), curSize)
 
 	// Open a Reader
 	reader, err := suite.StorageDriver.Reader(suite.ctx, filename, 0)
-	c.Assert(err, check.IsNil)
+	suite.Require().NoError(err)
 	defer reader.Close()
 
 	// Check the file is empty
 	buf := make([]byte, chunkSize)
 	n, err := reader.Read(buf)
-	c.Assert(err, check.Equals, io.EOF)
-	c.Assert(n, check.Equals, 0)
+	suite.Require().ErrorIs(err, io.EOF)
+	suite.Require().Equal(0, n)
 
 	// Open a Writer for Append
 	awriter, err := suite.StorageDriver.Writer(suite.ctx, filename, true)
-	c.Assert(err, check.IsNil)
+	suite.Require().NoError(err)
```

Comment:
`TestWriteZeroByteStreamThenAppend` is currently failing for S3 driver. The `filename` does exist but `suite.StorageDriver.Writer` returns `s3aws: Path not found` error.

Reply:
@Juneezee this looks to have this same root cause as the issue we resolved here: https://gitlab.com/gitlab-org/container-registry/-/merge_requests/20, namely:

> The registry S3 storage driver uses multipart uploads for streaming. Multipart uploads require at least 1 chunk, so it doesn’t work for 0B files (0 chunks). 

The same is true for instantiating the writer when append is set to `true`.

Reply:
The GitLab container registry "resolved" it by skipping a benchmark. This is a unit test. All the other storage drivers pass this test. Code which consumes the storage driver interface may therefore depend on the behaviour asserted by this test. The s3 driver is not upholding the contract, therefore the driver contains a bug which needs fixing.

Reply:
I agree, S3 should conform to expectations here, the linked MR demonstrates a likely root cause, even if it wasn't addressed at the time.

Reply:
I'm only catching up on this now, but I almost feel like this test might never work with S3 driver as is.

I'm talk abuot about `TestWriteZeroByteContentThenAppend`. Specifically what it does is:
1. It _PUTs_ the [empty] content into S3 bucket https://github.com/distribution/distribution/blob/04e3bdaa7c6304d14970f13c01f0a8ded8289e8d/registry/storage/driver/testsuites/testsuites.go#L551
2. Tries to _LIST MULTIPART_ objects

The reason I'm highlighting the _PUT_ and _LIST MULTIPART_ is that IIRC `PutContent` smacks `PUT /{Bucket}/{Key+}` and _LIST MULTIPART_ calls `GET /{Bucket}?uploads`. I dont think this is gonna fly because `PutContent` doesn't actually create a multipart upload.

What should work is if instead of `PutContent` S3 driver call we create a `Writer` in a non-`appendMode` and `Write` empty content which should actually create a multipart upload which can then be listed and appended to. 

So whilst I'm _not_ ruling out the existence of a bug in the `S3` driver, I'm feeling fairly confident this test will NOT work as is in the S3 driver 🤷‍♂️

Reply:
I've played around with this on my personal AWS account and it has revived an unpleasant AWS S3 memory for me 🫠

When you initiate a [multipart upload](https://docs.aws.amazon.com/AmazonS3/latest/API/API_MultipartUpload.html) to S3, it first creates an upload "record" but does not actually upload any object data. Those are uploaded as parts. So for an empty 0 byte file, no object content is uploaded, just the metadata for the initiated upload record is stored. We actually handle this edge case in the S3 driver by uploading an empty part. I'm wondering if that's actually the problem here, but I'm not sure 🤷‍♂️ 

What's happening in`TestWriteZeroByteStreamThenAppend` test is:
1. We create a multipart upload when the non-`appendMode` writer is created: https://github.com/distribution/distribution/blob/04e3bdaa7c6304d14970f13c01f0a8ded8289e8d/registry/storage/driver/s3-aws/s3.go#L685-L695
2. We then commit and write the empty part (`Commit` and `Close` calls actually lead to this call https://github.com/distribution/distribution/blob/04e3bdaa7c6304d14970f13c01f0a8ded8289e8d/registry/storage/driver/s3-aws/s3.go#L1583)
3. We then try to list multipart uploads when we create the `appendMode` writer here: https://github.com/distribution/distribution/blob/04e3bdaa7c6304d14970f13c01f0a8ded8289e8d/registry/storage/driver/s3-aws/s3.go#L707 That call returns no `Uploads` because there are no pending uploads since the one we had created earlier has been successfully completed via the `Commit`. The `Write` call actually `break`s here: https://github.com/distribution/distribution/blob/04e3bdaa7c6304d14970f13c01f0a8ded8289e8d/registry/storage/driver/s3-aws/s3.go#L715 which then ends up with the Not Found path error returned here: https://github.com/distribution/distribution/blob/04e3bdaa7c6304d14970f13c01f0a8ded8289e8d/registry/storage/driver/s3-aws/s3.go#L758

I feel like I'm missing some S3 multipart uploads knowledge despite trawling through their docs. I wrote a simple Go program that verifies that completed uploads are not returned in the list multipart uploads, only the ongoing ones:
[see this gist](https://gist.github.com/milosgajdos/2274ad8999d978697ad7da5da7f77275) (just remove the `svc.CompleteMultipartUploadWithContext` call and list and you'll see your pending upload returned in the list call -- make you you change the key to something that's not uploaded yet)

If this is really the correct behaviour then I struggle to see how the `TestWriteZeroByteStreamThenAppend` can ever pass 🤔 I also might be overlooking something, so more heads looking into this would be helpful 😄

Reply:
Hmm, and in the case of `TestWriteZeroByteContentThenAppend` it's actually kinda similar but not. The test doesn't even do `MultipartUpload` -- it calls a `PutContent` call:
https://github.com/distribution/distribution/blob/e77c81c50b1fba53e4c6d32d641e0d55246f6664/registry/storage/driver/testsuites/testsuites.go#L532

Which does a `PutObject`: https://github.com/distribution/distribution/blob/e77c81c50b1fba53e4c6d32d641e0d55246f6664/registry/storage/driver/s3-aws/s3.go#L649-L661

Then the test creates the `appendMode` writer https://github.com/distribution/distribution/blob/e77c81c50b1fba53e4c6d32d641e0d55246f6664/registry/storage/driver/testsuites/testsuites.go#L547 which as I said above attempts listing Multipart uploads - that ain't gonna work, no, I don't think so.

Now, both these tests pass with GCS and Azure drivers (I've verified it on my personal accounts - happy to do a screen recording if need be), but S3 driver might need massaging as I don't believe S3 supports appending data to existing keys other than via multipart uploads...

Reply:
> Thinking through this, is there a case in the code where we open a writer without append and then later append to the file? 

I don't think so, no. I haven't been able to find any such occurrences.

> When using the writer implementation in the blobwriter, we only disable append when we're acting as a proxy, and then we're not going to be using any of the cloud drivers, right?

Are you saying that since this is an implementation detail, it might not be worth testing this behaviour?

> The s3aws: Path not found from this test is probably an example of the latter.

This is actually our own doing 😄  we massage the error into this shape, but V1 SDK is not giving us much choice. Point in case: https://github.com/distribution/distribution/blob/bdf70a1e462dad2f441ad2331a9352ce45ce5c5c/registry/storage/driver/s3-aws/s3.go#L1237

Anyway, this is a bit of a pickle that makes me wonder if this test ever passed before. I've tried running the integration tests on some old commits way back and found them failing on this test (and a few others, actually, so getting to this point is an improvement -- just trying to put a positive spin on this 😄 ).

I also think it might be worth revisiting the `gcs` and `azure` drivers again and understand how this works there alas very different implementation semantics, though the high level concepts could give us some inspo...I really don't think this test can ever pass. 

One **very silly HACK** we could do is, read the object in and "reupload" it as a multipart upload with the new data appended to it

Reply:
> Are you saying that since this is an implementation detail, it might not be worth testing this behaviour?

The more tightly we define the behaviors of the driver interface via these test suites, the more hacky stuff we'll need to do in the driver implementations to make them conform. The driver code is complex enough as it is already, so if we don't have a solid case for needing this behavior, we should reconsider enforcing it.

> Anyway, this is a bit of a pickle that makes me wonder if this test ever passed before. I've tried running the integration tests on some old commits way back and found them failing on this test (and a few others, actually, so getting to this point is an improvement -- just trying to put a positive spin on this 😄 ).

To my point above, if this has worked in the wild for a few years with this test failing, that weakens the case for this test existing.

Reply:
> To my point above, if this has worked in the wild for a few years with this test failing, that weakens the case for this test existing.

I tend to agree with this, but I'd also say we should try to find a way to make this work _if possible_ and if not, we should document it -- at the very least in the `godoc` comment.

What's your take on this @corhere

Reply:
Getting the S3 driver to conform to the interface by narrowing the interface contract is fine with me. Emphasis on interface --- not documenting caveats on the driver implementation. For example, we could update the `Writer` doc to say that only uncommitted paths can be appended to.

Though given that both of the failing tests have "write zero bytes to a path" as a precondition, how hacky would it really be to modify the S3 driver to handle appending to an empty object by overwriting it?

Reply:
> how hacky would it really be to modify the S3 driver to handle appending to an empty object by overwriting it?

This is easy. The question is whether we should allow this behaviour only for the blobs with 0-byte size or if it should be allowed in general for any object or "arbitrary" size. The latter might lead to some unpleasant situations hence my calling it a hack, though if we document it properly end users could be warned

Reply:
Given how memory- and bandwidth-intensive it would be to append to an object with a read-modify-write cycle, and that it is not needed by the current app, I don't see the need to have the S3 driver support it. Application code could implement that behaviour itself through the abstract driver interface—by reading then overwriting the path—if the need ever arose. (On that note, it may be a smart move to document whether concurrently writing to and reading from the same path is permitted.)

Do you think the following interface doc changes would be sufficient?
```diff
diff --git a/registry/storage/driver/storagedriver.go b/registry/storage/driver/storagedriver.go
index d8464576..efb68bd0 100644
--- a/registry/storage/driver/storagedriver.go
+++ b/registry/storage/driver/storagedriver.go
@@ -59,11 +59,12 @@ type StorageDriver interface {
 	// name, but drivers may provide other information here.
 	Name() string
 
-	// GetContent retrieves the content stored at "path" as a []byte.
+	// GetContent retrieves the content previously committed to "path".
 	// This should primarily be used for small objects.
 	GetContent(ctx context.Context, path string) ([]byte, error)
 
-	// PutContent stores the []byte content at a location designated by "path".
+	// PutContent commits the content to the location designated by "path",
+	// replacing any existing content.
 	// This should primarily be used for small objects.
 	PutContent(ctx context.Context, path string, content []byte) error
 
@@ -74,6 +75,11 @@ type StorageDriver interface {
 
 	// Writer returns a FileWriter which will store the content written to it
 	// at the location designated by "path" after the call to Commit.
+	// A path may be appended to if it has not been committed, or if the
+	// existing committed content is zero length.
+	//
+	// The behaviour of appending to paths with non-empty committed content is
+	// undefined. Specific implementations may document their own behavior.
 	Writer(ctx context.Context, path string, append bool) (FileWriter, error)
 
 	// Stat retrieves the FileInfo for the given path, including the current
```
I lifted the "behavior...is undefined" phrasing from the [`io.Closer`](https://pkg.go.dev/io#Closer) doc.

Reply:
I prefer sticking to `stored content` as it can be less confusing to the end users reading the docs. `PutContent` actually doesn't commit (meaning it doesn't call `(writer).Commit`) so that seems a bit misleading 🤔

Unless you meant something else by _commit_. I do like the phrasing of append, though -- calling it undefined and/or deferring to individual driver implementation is the right move IMHO.

Reply:
I figured. I should have been clear what my point was about: since `FileWriter` has a `Commit` method as you mention, saying "PutContent commits the content" might make some people incorrectly assume that `PutContent` actually calls `(FileWriter).Commit()`. That said, there are so many overloaded terms 😅

Reply:
"PutContent stores and commits the content"?

Reply:
As for the S3 driver implementation, something like this could do the trick
```diff
--- registry/storage/driver/s3-aws/s3.go
+++ registry/storage/driver/s3-aws/s3.go
@@ -713,7 +713,27 @@ func (d *driver) Writer(ctx context.Context, path string, appendMode bool) (stor
                // if there were no more results to return after the first call, resp.IsTruncated would have been false
                // and the loop would be exited without recalling ListMultipartUploads
                if len(resp.Uploads) == 0 {
-                       break
+                       fi, err := d.Stat(ctx, path)
+                       if err != nil {
+                               return nil, parseError(path, err)
+                       }
+
+                       if fi.Size() == 0 {
+                               resp, err := d.S3.CreateMultipartUploadWithContext(ctx, &s3.CreateMultipartUploadInput{
+                                       Bucket:               aws.String(d.Bucket),
+                                       Key:                  aws.String(key),
+                                       ContentType:          d.getContentType(),
+                                       ACL:                  d.getACL(),
+                                       ServerSideEncryption: d.getEncryptionMode(),
+                                       SSEKMSKeyId:          d.getSSEKMSKeyID(),
+                                       StorageClass:         d.getStorageClass(),
+                               })
+                               if err != nil {
+                                       return nil, err
+                               }
+                               return d.newWriter(ctx, key, *resp.UploadId, nil), nil
+                       }
+                       return nil, storagedriver.ErrUnsupportedMethod{DriverName: driverName}
```

2 things:
1. I haven't tested this (will do later on)
2. `CreateMultiPartUpload` code could be extracted to a func (it's a copy-pasta from the case where we don't use append writer); implementation detail.
3. Returning `ErrUnsupportedMethod` does not feel quite right, though maybe it's ok to use it there; alternatively we could create a new error in `storagedriver` package something like `ErrUndefinedBehaviour` 🤷‍♂️

Reply:
> Returning `ErrUnsupportedMethod` does not feel quite right, though maybe it's ok to use it there

I only see `ErrUnsupportedMethod` when the entire method is not implemented, like the `URLFor` method in the filesystem driver, so I don't think we should use it here.

> alternatively we could create a new error in `storagedriver` package something like `ErrUndefinedBehaviour`

How do you feel about returning a generic error for now? Unless there are other places where `ErrUndefinedBehaviour` could live.

Reply:
> I only see ErrUnsupportedMethod when the entire method is not implemented, like the URLFor method in the filesystem driver, so I don't think we should use it here.

Fair.

> How do you feel about returning a generic error for now? Unless there are other places where ErrUndefinedBehaviour could live.

No strong feelings: it'd be nice if storage drivers returned context-specific errors, but in this case it's...complicated I  suppose. I can't think of a project where I got an error that explicitly calls out the cause as undefined -- it's all been mostly random errors that never helped me solve the problems and, in fact, created more problems for me 😄

FYI: With the diff I pasted above, all tests pass. Can push to this branch once we've agreed on what we should return.

Reply:
> With the diff I pasted above, all tests pass. Can push to this branch once we've agreed on what we should return.

Nice! I would argue that the behavior of appending after commit is implementation dependent rather than "true" undefined behavior, so I'm in favor of a generic error here.

Reply:
Went with this: https://github.com/distribution/distribution/pull/4185/commits/4baddbc6087442615ec0dff1027c6ccef1a4cb27

Reply:
That looks good except for the comment mentioning `ErrUnsupportedMethod`: https://github.com/distribution/distribution/commit/4baddbc6087442615ec0dff1027c6ccef1a4cb27#diff-42ab115e8f2c28fa837e174c081ffe3542e13f5fbd58271f06b179c1123c7905R685

Reply:
Arrrrgh, that was in the original code I never cleaned up after we discussed it. Ok should be fixed now https://github.com/distribution/distribution/pull/4185/commits/3f3e61e2998454a6315ea32f55a2a52ad8d227f9

Reply:
Thanks, this LGTM :rocket: